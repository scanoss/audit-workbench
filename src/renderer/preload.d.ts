import {IpcRendererEvent} from "electron";
import {IpcChannels} from "@api/ipc-channels";

declare global {
  interface Window {
    electron: {
      ipcRenderer: {
        send(channel: IpcChannels, ...args: unknown[]): void;
        on(channel: IpcChannels, func: (_event: IpcRendererEvent, ...args: unknown[]) => void): (() => void) | undefined;
        once(channel: IpcChannels, func: (_event: IpcRendererEvent, ...args: unknown[]) => void): void;
        invoke(channel: IpcChannels, ...args: unknown[]): Promise<any>;
        removeListener(channel: IpcChannels, func: (...args: any) => void): void;
      };
    };
    os: {
      homedir(): string;
    };
    shell: {
      showItemInFolder(path: string): void;
      openExternal(url: string, options?: Electron.OpenExternalOptions): Promise<void>;
    };
    path: {
      resolve(p: string, s: string): string;
      sep: string;
    };
  }
}

export {};
