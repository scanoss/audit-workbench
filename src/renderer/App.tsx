import React from 'react';
import { createRoot } from 'react-dom/client';
import { HashRouter, Route, Routes } from 'react-router-dom';
import { Provider } from 'react-redux';

import AppConfig from '@config/AppConfigModule';
import { DialogProvider } from '@context/DialogProvider';
import { WorkbenchProvider } from '@context/WorkbenchProvider';
import AppProvider from '@context/AppProvider';
import store from '@store/store';
import { adaptV4Theme, createMuiTheme, StyledEngineProvider, Theme } from '@mui/material';
import WorkbenchModule from './features/workbench';
import WorkspaceModule from './features/workspace';
import AboutModule from './features/about';

import './App.global.scss';
import { ThemeProvider } from '@mui/styles';


export default class App {
  /**
   * Initialize React application.
   *
   * @returns {JSX.Element}
   */
  public async setup(): Promise<void | Element | React.Component> {
    this.setTitle();
    const theme = this.loadTheme();

    const app = (
      <>
        <HashRouter>
          <Provider store={store}>
            <StyledEngineProvider injectFirst>
              <ThemeProvider theme={theme}>
                <DialogProvider>
                  <AppProvider>
                    <WorkbenchProvider>
                      <Routes>
                        <Route index element={<WorkspaceModule />} />
                        <Route path="/workspace/*" element={<WorkspaceModule />} />
                        <Route path="/workbench/*" element={<WorkbenchModule />} />
                        <Route path="/about" element={<AboutModule />} />
                      </Routes>
                    </WorkbenchProvider>
                  </AppProvider>
                </DialogProvider>
              </ThemeProvider>
            </StyledEngineProvider>
          </Provider>
        </HashRouter>
      </>
    );

    this.setupAppMenuListeners();

    const container = document.getElementById('root')!;
    const root = createRoot(container);
    root.render(app);
  }

  private setTitle() {
    document.title = AppConfig.APP_NAME;
  }

  setupAppMenuListeners() {
    /* ipcRenderer.on(IpcEvents.MENU_OPEN_SETTINGS, async (event) => {
      console.log('menu open', event);
      // window.location.hash = '/workbench/detected';
    }); */
  }

  private loadTheme(): Theme {
    const theme = createMuiTheme(adaptV4Theme({
      palette: {
        primary: {
          main: '#6366F1',
        },
        secondary: {
          main: '#22C55E',
          contrastText: '#FFFFFF',
        },
        success: {
          main: '#4caf50',
        },
      },
      typography: {
        button: {
          fontWeight: 600,
          textTransform: 'none',
        },
        fontFamily: '"Inter", "Roboto", "Helvetica", "Arial", sans-serif',
      },
    }));

    theme.shadows[1] = '0px 1px 3px 0px #0000001A; 1px 0px 2px 0px #0000000F';

    return theme;
  }
}
