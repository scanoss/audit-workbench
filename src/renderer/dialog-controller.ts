const electron = window.require('electron');
const { remote } = electron;
const { dialog } = remote;

class DialogController {
  public showOpenDialog(options): string {
    const result = dialog.showOpenDialogSync(remote.getCurrentWindow(), options);
    return result ? result[0] : null;
  }

  public showError(title: string, content: string) {
    dialog.showErrorBox(title, content);
  }
}

export const dialogController = new DialogController();
