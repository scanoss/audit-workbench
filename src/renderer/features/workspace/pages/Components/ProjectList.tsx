import React from 'react';
import {
  IconButton,
  Link,
  makeStyles,
  Paper,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Tooltip,
  Chip,
} from '@material-ui/core';
import DeleteIcon from '@material-ui/icons/Delete';
import ReplayIcon from '@material-ui/icons/Replay';
import PlayArrowIcon from '@material-ui/icons/PlayArrow';
import WarningOutlinedIcon from '@material-ui/icons/WarningOutlined';
import GetAppOutlined from '@material-ui/icons/GetAppOutlined';
import { IProject, ScanState } from '../../../../../api/types';
import AppConfig from '../../../../../config/AppConfigModule';

const filter = (items, query) => {
  if (!items) return null;

  const result = items.filter((item) => {
    const name = item.name.toLowerCase();
    return name.includes(query.toLowerCase());
  });

  return result;
};

const format = (date) => {
  return new Date(date).toLocaleDateString('en-US', {
    day: 'numeric',
    month: 'short',
    year: 'numeric',
  });
};

const useStyles = makeStyles((theme) => ({
  md: {
    maxWidth: 130,
    textAlign: 'center',
  },
}));

const isProjectFinished = (project: IProject): boolean => project.scannerState === ScanState.FINISHED;
const isProjectDeprecated = (project: IProject): boolean => project.appVersion < AppConfig.MIN_VERSION_SUPPORTED;
const isProjectImported = (project: IProject): boolean => project.source === 'IMPORTED';

interface ProjectListProps {
  projects: IProject[];
  searchQuery: string;
  onProjectClick: (project: IProject) => void;
  onProjectDelete: (project: IProject) => void;
  onProjectRestore: (project: IProject) => void;
  onProjectRescan: (project: IProject) => void;
  onProjectExport: (project: IProject) => void;
  onProjectCreate: () => void;
  onProjectImport: () => void;
}

const ProjectList = (props: ProjectListProps) => {
  const classes = useStyles();

  const { projects, searchQuery } = props;
  const filterProjects = filter(projects, searchQuery);

  return (
    <>
      {projects && projects.length > 0 ? (
        <TableContainer component={Paper}>
          <Table className="projects-table" aria-label="projects table">
            <TableHead>
              <TableRow>
                <TableCell width="50%">NAME</TableCell>
                <TableCell>DATE</TableCell>
                <TableCell>TOTAL FILES</TableCell>
                <TableCell width={30} />
              </TableRow>
            </TableHead>
            <TableBody>
              {filterProjects.length !== 0 ? (
                filterProjects.map((project) => (
                  <TableRow
                    className={`
                      ${isProjectFinished(project) ? 'scanning-complete' : 'scanning-not-complete'}
                      ${isProjectDeprecated(project) ? 'project-deprecated' : ''}
                      `}
                    hover
                    key={project.name}
                    onClick={() => isProjectFinished(project) && props.onProjectClick(project)}
                  >
                    <TableCell component="th" scope="row">
                      <div className="project-name">
                        {isProjectDeprecated(project) && (
                          <Tooltip
                            classes={{ tooltip: classes.md }}
                            title="This project was scanned with a previous version that is no longer supported."
                          >
                            <WarningOutlinedIcon fontSize="inherit" className="icon mr-1" />
                          </Tooltip>
                        )}
                        <span>{project.name}</span>
                        {isProjectImported(project) && (
                          <Tooltip classes={{ tooltip: classes.md }} title="This project was imported. You will not be able to see the original source code.">
                            <Chip label="IMPORTED" size="small" variant="outlined" className="ml-1" />
                          </Tooltip>
                        )}
                      </div>
                    </TableCell>
                    <TableCell>{format(project.date)}</TableCell>
                    <TableCell>{project.files}</TableCell>
                    <TableCell className="row-actions">
                      <div className="btn-actions">
                        {!isProjectFinished(project) && !isProjectDeprecated(project) && (
                          <Tooltip title="Resume scan">
                            <IconButton
                              aria-label="restore"
                              className="btn-restore"
                              onClick={(event) => {
                                event.stopPropagation();
                                props.onProjectRestore(project);
                              }}
                            >
                              <PlayArrowIcon fontSize="small" />
                            </IconButton>
                          </Tooltip>
                        )}

                        {isProjectFinished(project) && !isProjectDeprecated(project) && (
                          <>
                          <Tooltip title="Export project">
                              <IconButton
                                aria-label="export"
                                className="btn-export"
                                onClick={(event) => {
                                  event.stopPropagation();
                                  props.onProjectExport(project);
                                }}
                              >
                                <GetAppOutlined fontSize="small" />
                              </IconButton>
                            </Tooltip>

                            <Tooltip title="Rescan">
                              <IconButton
                                aria-label="rescan"
                                className="btn-rescan"
                                onClick={(event) => {
                                  event.stopPropagation();
                                  props.onProjectRescan(project);
                                }}
                              >
                                <ReplayIcon fontSize="small" />
                              </IconButton>
                            </Tooltip>
                          </>
                        )}

                        <Tooltip title="Remove project">
                          <IconButton
                            aria-label="delete"
                            className="btn-delete"
                            onClick={(event) => {
                              event.stopPropagation();
                              props.onProjectDelete(project);
                            }}
                          >
                            <DeleteIcon fontSize="small" />
                          </IconButton>
                        </Tooltip>
                      </div>
                    </TableCell>
                  </TableRow>
                ))
              ) : (
                <TableRow>
                  <TableCell colSpan={4}>
                    <p className="text-center">
                      Not projects found with <strong>{searchQuery}</strong>
                    </p>
                  </TableCell>
                </TableRow>
              )}
            </TableBody>
          </Table>
        </TableContainer>
      ) : !projects ? (
        <p>Loading projects...</p>
      ) : (
        <div className="empty-container">
          <div className="empty-list">
            <h3>No projects found</h3>
            <p>
              You can start by&nbsp;
              <Link onClick={() => props.onProjectCreate()}>scanning a new project</Link><br />or <Link onClick={() => props.onProjectImport()}>importing a project</Link>.
            </p>
          </div>
        </div>
      )}
    </>
  );
};

export default ProjectList;
