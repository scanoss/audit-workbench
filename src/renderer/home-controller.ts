import { projectService } from '../api/project-service';
import { workspaceService } from '../api/workspace-service';
import { IpcEvents } from '../ipc-events';

const { ipcRenderer } = require('electron');

export const scan = (path: string) => {
  ipcRenderer.send(IpcEvents.SCANNER_INIT_SCAN, { path });
};

export const resume = async (path: string) => {
  const response = await projectService.resume(path);
};

export const rescan = async (path: string) => {
  await workspaceService.rescanProject(path);
};

export const open = async (path: string) => {
   const response = await projectService.load(path);
   return response;
};

export const defaultWorkspacePath = async () => {
  const response = await projectService.workspacePath();
  return response;
};
