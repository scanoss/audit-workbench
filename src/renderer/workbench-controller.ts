import { projectService } from '../api/project-service';
import { componentService } from '../api/component-service';
import { Component } from '../api/types';

const fs = require('original-fs').promises;

export interface ScanResult {
  scan: Record<string, unknown>;
  scanRoot: string;
  fileTree: any[];
}

class WorkbenchController {
  /**
   * Open scan file result
   *
   * @param {string} path
   * @returns {Promise<ScanResult>}
   * @memberof WorkbenchController
   */
  public async loadScan(path: string): Promise<ScanResult> {
    const { data } = await projectService.load(path);
    console.log(data);

    return this.generateScanResult(data);
  }

  /**
   * Get file content from a local file
   *
   * @param {string} path
   * @returns {string}
   * @memberof WorkbenchController
   */
  public async fetchLocalFile(path: string): Promise<string> {
    const data = await fs.readFile(path, 'utf-8');
    return data;
  }

  /**
   * Get file content from a remote file
   *
   * @param {string} hash
   * @returns {string}
   * @memberof WorkbenchController
   */
  public async fetchRemoteFile(hash: string): Promise<string> {
    // TODO: move api url to API Config
    const response = await fetch(`https://osskb.org/api/file_contents/${hash}`);
    if (!response.ok) throw new Error('File not found');

    return response.text();
  }

  public async getComponents(): Promise<Component[]> {
    const response = await componentService.getAll({});
    console.log('COMPONENTS', response.data);
    return response.data;
  }

  public async getComponent(id: number): Promise<Component> {
    const response = await componentService.get({compid: id});
    console.log('COMPONENT', response.data);
    return response.data;
  }

  private async generateScanResult(data): Promise<ScanResult> {
    const scan = data.results;
    return {
      scan,
      scanRoot: data.scan_root,
      fileTree: [data.logical_tree],
    };
  }
}

export const workbenchController = new WorkbenchController();
