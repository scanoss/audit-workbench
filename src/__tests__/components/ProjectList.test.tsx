import React from 'react';
import { configure } from '@testing-library/react';
import renderer from 'react-test-renderer';
import ProjectList from '../../renderer/features/workspace/pages/Components/ProjectList';

const props = {
  projects: [],
  searchQuery: '',
  onProjectClick: jest.fn(),
  onProjectDelete: jest.fn(),
  onProjectRestore: jest.fn(),
  onProjectCreate: jest.fn(),
  onProjectImport: jest.fn(),
  onProjectExport: jest.fn(),
  onProjectRescan: jest.fn(),
};

describe('ProjectList', () => {
  it('should render an empty container', () => {
    const component = renderer.create(<ProjectList {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
