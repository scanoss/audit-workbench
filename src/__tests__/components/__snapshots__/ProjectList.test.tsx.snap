// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectList should render an empty container 1`] = `
<div
  className="empty-container"
>
  <div
    className="empty-list"
  >
    <h3>
      No projects found
    </h3>
    <p>
      You can start byÂ 
      <a
        className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover css-17dcqu8-MuiTypography-root-MuiLink-root"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
      >
        scanning a new project
      </a>
      <br />
      or
       
      <a
        className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover css-17dcqu8-MuiTypography-root-MuiLink-root"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
      >
        importing a project
      </a>
      .
    </p>
  </div>
</div>
`;
